server/src/main/java/org/opensearch/index/mapper/DocumentParser.java

```diff
@@ -1010,4 +1040,51 @@ private static Mapper getMapper(final ParseContext context, ObjectMapper objectM
         }
         return objectMapper.getMapper(subfields[subfields.length - 1]);
     }
+
+    @SuppressWarnings("rawtypes")
+    private static Mapper.Builder findTemplateBuilder(
```

Comment:
So, I think we need to spend a bit more time on how implement `findTemplateBuilder` here: all two methods do the same thing - they do call `findTemplate` and return `null` (or continue). I think we could keep the old methods `RootObjectMapper` without any changes, but only keep `findTemplate` method here which will do check for dynamic part. Thoughts?

Reply:
Thanks, I've added the two methods in `RootObjectMapper` back and add a new method `findTemplateBuilderByTemplate` in `RootObjectMapper`, we need to call that new method here in `DocumentParser` because we already get the template by the `findTemplate` method.

Reply:
But why do we need this `findTemplate` ? Couldn't we have a single method here:

```
private static Mapper.Builder findTemplateBuilder(
        ParseContext context,
        String name,
        XContentFieldType matchType,
        DateFormatter dateFormat,
        ObjectMapper.Dynamic dynamic,
        String fieldFullPath
    ) {
      Mapper.Builder builder = context.root().findTemplateBuilder(context.path(), name, matchType);
      
       if (builder == null && dynamic == ObjectMapper.Dynamic.STRICT_ALLOW_TEMPLATES) {
            throw new StrictDynamicMappingException(dynamic.toString(), fieldFullPath, name);
        }

        return builder;
}
```

We know that the only condition when `findTemplateBuilder` is because the dynamic template is `null`.

Reply:
We need two `findTemplateBuilder` methods because one is for date field with `DateFormatter` and another is for all non-date fields, they have different parameters.

Reply:
OK, I got your point, now no changes happen in `RootObjectMapper`, we call the two public `RootObjectMapper::findTemplateBuilder` methods here.